Assets/Scripts/Encounters/ShopEncounter.cs

```diff
@@ -27,69 +27,70 @@ public class KeepsakeInShopWithPrice {
 [System.Serializable]
 public class ShopEncounter : Encounter
 {
-    public bool generateEncounter;
     public ShopDataSO shopData;
     public List<CardInShopWithPrice> cardsInShop = new List<CardInShopWithPrice>();
     public List<KeepsakeInShopWithPrice> keepsakesInShop = new List<KeepsakeInShopWithPrice>();
     private EncounterConstantsSO encounterConstants;
 
+    private ShopUIManager shopUIManager;
+
     public ShopEncounter() {
         this.encounterType = EncounterType.Shop;
     }
 
     public ShopEncounter(ShopDataSO shopData) {
         this.encounterType = EncounterType.Shop;
         this.shopData = shopData;
-        generateShopEncounter();
     }
 
     public override void build(List<Companion> companionList, EncounterConstantsSO constants)
     {
+        if (shopUIManager == default) {
+            findShopSection();
+        }
+
         this.encounterType = EncounterType.Shop;
         ShopManager.Instance.saveShopEncounter(this);
         this.encounterConstants = constants;
-        if (generateEncounter) {
-            generateShopEncounter();
-        }
-        setupCards();
+        
+        generateShopEncounter(companionList);
+        setupCards(companionList);
         setupKeepsakes();
         this.shopData.shopEncounterEvent.Raise(this);
     }
 
-    private void setupCards() {
-        if (cardsInShop.Count > shopData.cardLocations.Count) {
-            Debug.LogError("Unable to setup cards in shop encounter, not enough locations for cards");
-            return;
-        }
-        
+    private void findShopSection() {
+        shopUIManager = GameObject.FindObjectOfType<ShopUIManager>();
```

Comment:
Not sure if I added the Singleton code to main yet, but should use that here when we do